ros2/ament.bzl

```diff
@@ -390,3 +391,99 @@ def sh_launcher(name, deps, idl_deps = None, **kwargs):
         ament_setup = ament_setup,
         **kwargs
     )
+
+def _cpp_ament_setup_impl(ctx):
+    src_output = ctx.actions.declare_file(ctx.attr.basename + "/ament_setup.cc")
+    hdr_output = ctx.actions.declare_file(ctx.attr.basename + "/ament_setup.h")
+    ament_prefix_path = ctx.attr.ament_setup[Ros2AmentSetupInfo].ament_prefix_path
+
+    substitutions = {
+        "{{ament_prefix_path}}": ament_prefix_path,
+        "{{header}}": hdr_output.short_path,
+        "{{namespace}}": ctx.attr.basename,
+    }
+
+    expand_template_impl(
+        ctx,
+        template = ctx.file._src_template,
+        output = src_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+    expand_template_impl(
+        ctx,
+        template = ctx.file._hdr_template,
+        output = hdr_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+
+    files = depset([src_output, hdr_output])
+    return [
+        DefaultInfo(
+            files = files,
+        ),
+    ]
+
+cpp_ament_setup = rule(
+    attrs = {
+        "ament_setup": attr.label(
+            mandatory = True,
+            providers = [Ros2AmentSetupInfo],
+        ),
+        "basename": attr.string(
```

Comment:
OK, we can stick with basename. It doesn't need to be mandatory IMO. You can do something like in the impl:
```py
basename = ctx.attr.basename or ctx.attr.name
src_output = ctx.actions.declare_file(basename + "/ament_setup.cc")
...
```